### ISSUE TITLE
CycleGAN tranining pipeline in Documentation failing silently with ipython kernel dying

### ISSUE THREAD

**Issue by lzyang2000** [tensorflow/tensorflow#41474]
<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.6.9
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.1 /  7.6.4.38-1+cuda10.1
- GPU model and memory: GTX1080Ti / 12GB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When training a custom img2img cycleGAN on 4600+ images in each domain (9300+ images in total), the ipython kernel of the jupyter notebook dies silently some time into training.
**Describe the expected behavior**
Training proceeds normally.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
Due the data problem, it cannot really be reproduced elsewhere.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The one log I find fishy:

```
 The calling iterator did not fully read the dataset being cached. In order to avoid unexpected truncation of the dataset, the
 partially cached contents of the dataset will be discarded. This can happen if you have an input pipeline similar to 
`dataset.cache().take(k).repeat()`. You should use `dataset.take(k).cache().repeat()` instead.
```
This is the code I have:
```
train_A = train_A.map(
    preprocess_image_train, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)
```
I think I should change it to:
```
train_A = train_A.map(
    preprocess_image_train, num_parallel_calls=AUTOTUNE).batch(1).cache().shuffle(
    BUFFER_SIZE)
```
Is that correct?

Thank you!

**Comment by amahendrakar** [comment]
@lzyang2000,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here or share the link to the documentation you're following. Thanks!

**Comment by lzyang2000** [comment]
Here is the link to the documentation that I am following:
[https://www.tensorflow.org/tutorials/generative/cyclegan](https://www.tensorflow.org/tutorials/generative/cyclegan)
It is the same as my code save for the difference in the amount of data loaded.

Thank you!

**Comment by amahendrakar** [comment]
@lzyang2000,
I was able to run the code without any issues, please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/96cff7fa952064730857a281591c0568/41474.ipynb). 

Could you try running the code in a virtual environment and check if you are facing the same issue. Thanks!

**Comment by lzyang2000** [comment]
Sorry for the late response, yes the original tutorial code works, it's when I try to load a custom dataset with more images (5000+) then it fails.

The problem is I do not have the resources to scale up the memory of the server that I am working on, so I cannot confirm whether or not it's a memory issue.

Here is the code that I used to load the images:
`train_rgb = tf.data.Dataset.list_files("/tf/gan/cycleGAN/images/"+dataset_name+"/trainA/*.jpg")`
`train_rgb = train_rgb.map(
    preprocess_image_train, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)`

where `preprocess_image_train` is the same as in the original code.

**Comment by jvishnuvardhan** [comment]
@lzyang2000 Can you please share a standalone code to reproduce the issue? You don't need to share your private data as I can use available public data to reproduce the issue. Thanks!

**Comment by lzyang2000** [comment]
Thanks, here is the code:
```
import tensorflow as tf
dataset_name="a"
train_rgb = tf.data.Dataset.list_files("/tf/gan/cycleGAN/images/"+dataset_name+"/trainA/*.jpg")  # doctest: +SKIP
train_thm = tf.data.Dataset.list_files("/tf/gan/cycleGAN/images/"+dataset_name+"/trainB/*.jpg")  # doctest: +SKIP
test_rgb = tf.data.Dataset.list_files("/tf/gan/cycleGAN/images/"+dataset_name+"/testA/*.jpg")  # doctest: +SKIP
test_thm = tf.data.Dataset.list_files("/tf/gan/cycleGAN/images/"+dataset_name+"/testB/*.jpg")  # doctest: +SKIP
BUFFER_SIZE = 1000
BATCH_SIZE = 1
IMG_WIDTH = 1024
IMG_HEIGHT = 1024
def random_crop(image):
  cropped_image = tf.image.random_crop(
      image, size=[IMG_HEIGHT, IMG_WIDTH, 3])

  return cropped_image
# normalizing the images to [-1, 1]
def normalize(image):
  image = tf.cast(image, tf.float32)
  image = (image / 127.5) - 1
  return image
def random_jitter(image):
  # resizing to 286 x 286 x 3
  image = tf.image.resize(image, [1144, 1144],
                          method=tf.image.ResizeMethod.NEAREST_NEIGHBOR)

  # randomly cropping to 256 x 256 x 3
  image = random_crop(image)

  # random mirroring
  image = tf.image.random_flip_left_right(image)

  return image
def process_path(file_path):
  # load the raw data from the file as a string
  img = tf.io.read_file(file_path)
  img = decode_img(img)
  return img

def decode_img(img):
  # convert the compressed string to a 3D uint8 tensor
  img = tf.image.decode_jpeg(img, channels=3)
  # Use `convert_image_dtype` to convert to floats in the [0,1] range.
 # img = tf.image.convert_image_dtype(img, tf.float32)
  # resize the image to the desired size.
  return img

def preprocess_image_train(image,label=None):
  image = process_path(image)
  image = random_jitter(image)
  image = normalize(image)
  return image

def preprocess_image_test(image,label=None):
  image = process_path(image)
  image = normalize(image)
  return image

############# The place where TensorFlow logs the errors #############
train_rgb = train_rgb.map(
    preprocess_image_train, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)

train_thm = train_thm.map(
    preprocess_image_train, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)

test_rgb = test_rgb.map(
    preprocess_image_test, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)

test_thm = test_thm.map(
    preprocess_image_test, num_parallel_calls=AUTOTUNE).cache().shuffle(
    BUFFER_SIZE).batch(1)
######################################################################
```

**Comment by lzyang2000** [comment]
Hello, just wanting to see if there are any updates?

Thank you!

**Comment by jvishnuvardhan** [comment]
@lzyang2000 Sorry for late response. Can you please share a full standalone code? Can you try to use any public datasets such as `mnist` and make the code code a standalone?  Without standalone code, it is difficult to find root-cause of the error you are facing. thanks!

**Comment by jchwenger** [comment]
I should probably open a separate issue, but I have a feeling this might be related.

If you try and run the [CycleGAN tutorial](https://www.tensorflow.org/tutorials/generative/cyclegan), and you don't have a high-memory set-up it is likely to crash (especially if you wish to follow the next steps at the end and train for longer.

It seems to be a combination of the memory footprints of the matplotlib figures as well as the `train_step` function. Running the notebook as is, only changing the GPU to a V100 to speed things up, you can see the evolution of the RAM usage:

Just before running training, ~4GB of RAM used.
![Just before running training](https://github.com/tensorflow/tensorflow/assets/34098722/7e458ce3-064e-4bdf-aad0-fed865cef979)

First compilation of the `train_step` function, ~6.5GB.
![Screenshot from 2023-09-01 11-11-06](https://github.com/tensorflow/tensorflow/assets/34098722/d721fef3-1f28-42f5-9abf-1b0500a3f7be)

After the first epoch, first image display, ~7.6GB.
![Screenshot from 2023-09-01 11-14-28](https://github.com/tensorflow/tensorflow/assets/34098722/28ca2bcb-5028-4a8f-a45f-68bb73ce52ab)

Starting the third epoch, ~9.1GB.
![Screenshot from 2023-09-01 11-17-00](https://github.com/tensorflow/tensorflow/assets/34098722/b4185c90-d640-4275-8021-cd9b50ce33f9)

From what I can see, it just keeps growing linearly. I am looking into ways to make the memory footprint stable, but so far I haven't been able to (running the notebook without any matplotlib figures helps, but only delays the inevitable: after enough epochs, it'll run out of RAM).

I'm planning on using this for teaching purposes, it would be great if the memory usage was stable, as many students won't necessarily be up for paying for the higher-memory tier out of the box!

Any ideas on how to go about resolving that issue would be most welcome!
